src/MSBuildBinLogQuery/Token/TaskToken.cs

```diff
@@ -0,0 +1,27 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Build.Logging.Query.Token
+{
+    public class TaskToken : Token, IEquatable<TaskToken>
+    {
+        public TaskToken() : base()
+        {
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as TaskToken);
+        }
+
+        public bool Equals([AllowNull] TaskToken other)
+        {
+            return other != null;
+        }
+
+        public override int GetHashCode()
+        {
+            return 1;
```

Comment:
Is this going to cause problems if you have a `Dictionary<Token, _>`? I think everything would collide. Maybe instead switch to 
```suggestion
            return nameof(Token).GetHashCode();
```
? That would at least spread these out.

Reply:
Should it be `nameof(TaskToken).GetHashCode()`?

Reply:
Haven't read all the code or understood the rationale, yet, but IEquatable + inheritance is perilous IMHO.

Reply:
Done.

Reply:
Oh, I see it is only on the leaf nodes. Consider sealing anything with value semantics.

Reply:
Do you mean sealing all the leaf nodes here?